### ISSUE TITLE
AvoidEscapedUnicodeCharacters should support trailing '\' escape symbol

### ISSUE THREAD

**Issue by nmancus1** [checkstyle/checkstyle#9143]
https://checkstyle.org/config_misc.html#AvoidEscapedUnicodeCharacters

From https://github.com/checkstyle/checkstyle/pull/9135#discussion_r550064696:

>The docs says there is one more escape - for a trailing \. Something like \\$.

The above comment is referencing this doc: https://docs.oracle.com/javase/specs/jls/se15/html/jls-3.html#jls-3.10.7

_______________________________________________________

```
➜  src cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
        <module name="AvoidEscapedUnicodeCharacters">
            <property name="allowIfAllCharactersEscaped" value="true"/>
        </module>
    </module>
</module>

➜  src cat TestClass.java 
public class TestClass {
    static final String value1 = """
        \u03bc\
        \s\u03bc\
        """; // should be no violation on line 2
}

➜  src javac TestClass.java 

➜  src java -jar checkstyle-8.39-SNAPSHOT-all.jar -c config.xml TestClass.java
Starting audit...
[ERROR] src/TestClass.java:2:37: Unicode escape(s) usage should be avoided. [AvoidEscapedUnicodeCharacters]
Audit done.
Checkstyle ends with 1 errors.

```
AvoidEscapedUnicodeCharacters should not report a violation in this case, since all characters are escaped in this text block (`value1`).

**Comment by shashwatj07** [comment]
I don't think we understand the issue correctly.

Something to note:

```

>javac Test.java
// no output --- compiled successfully

>cat Test.java
public class Test {
   static final String value1 = """
   \u03bc
   \s\u03bc
   """;
}

>cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
        <module name="AvoidEscapedUnicodeCharacters">
            <property name="allowIfAllCharactersEscaped" value="true"/>
        </module>
    </module>
</module>

>java -jar checkstyle-8.41-all.jar -c config.xml Test.java
Starting audit...
[ERROR] C:\Users\Shashwat\Test.java:2:36: Unicode escape(s) usage should be avoided. [AvoidEscapedUnicodeCharacters]
Audit done.
Checkstyle ends with 1 errors.
```

I don't think our check is working as expected for text blocks. If it is correct, this issue doesn't make much sense since the example given above shall be a violation anyway.

Please correct me if I'm wrong. @romani @strkkk @pbludov @rnveach

**Comment by nmancus1** [comment]
This is a good find; I think we need to take a look at the pattern that we are using (`ALL_ESCAPED_CHARS`).  The problem seems to be the `\n` (line ending) in between escape characters:

```bash

➜  src cat MyClass.java
class MyClass {
    static final String value1 = """
            \u03bc
            \s\u03bc
            """;
    static final String value2 = """
            \s\u03bc
            """;
    static final String value3 = """
            \s\u03bc""";
    static final String value4 = """
            \u03bc
            \s\u03bc""";
    static final String value4 = """
            \u03bc\s\u03bc""";
}
➜  src java -jar checkstyle-8.41-all.jar -c config.xml MyClass.java
Starting audit...
[ERROR] /home/nick/Desktop/Tester/src/MyClass.java:2:37: Unicode escape(s) usage should be avoided. [AvoidEscapedUnicodeCharacters]
[ERROR] /home/nick/Desktop/Tester/src/MyClass.java:11:37: Unicode escape(s) usage should be avoided. [AvoidEscapedUnicodeCharacters]
Audit done.
Checkstyle ends with 2 errors.

```

**Comment by nmancus1** [comment]
Changed the name title of this issue; this check already supported the `\` line continuation escape.

**Comment by nmancus1** [comment]
@shashwatj07 I thought this was a simple oversight on my part during AvoidEscapedUnicodeCharacers text blocks update last summer, but this is not the case. If you would like to work on this issue, please do; I am closing #9712 .
